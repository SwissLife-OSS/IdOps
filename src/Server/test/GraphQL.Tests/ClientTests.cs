using System.Threading.Tasks;
using HotChocolate;
using HotChocolate.Execution;
using IdOps.Authorization;
using IdOps.Security;
using Snapshooter.Xunit;
using Squadron;
using Xunit;

namespace IdOps.GraphQL.Tests
{
    [Collection(TestCollectionNames.GraphQL)]
    public class ClientTests : TestHelper
    {
        public ClientTests(MongoResource resource) : base(resource)
        {
        }

        [Theory]
        [InlineData(Roles.Admin)]
        [InlineData(Roles.Edit)]
        public async Task ValidRequest(string role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .AddUser()
                .AddRole(role);
            
            await TestDataBuilder
                .New(Services)
                .SetupTenant()
                .ExecuteAsync();

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson().MatchSnapshot(x => x.IgnoreFields("**.id"));
        }

        [Theory]
        [InlineData(Roles.Read)]
        [InlineData(null)]
        public async Task CreateClient_ByUserWithoutPermission_IsDenied(string? role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .AddUser()
                .AddRole(role);

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson()
                .MatchSnapshot();
        }

        [Fact]
        public async Task CreateClient_NotAuthenticated()
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .SetAuthenticated(false);

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson()
                .MatchSnapshot();
        }
    }
}
