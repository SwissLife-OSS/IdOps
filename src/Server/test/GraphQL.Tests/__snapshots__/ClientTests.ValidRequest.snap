{
  "data": {
    "createClient": {
      "client": {
        "id": "cb7be06777bd4db38191df2b5b58f2bb",
        "name": "MyClient",
        "tenant": "TestTenant",
        "allowedGrantTypes": [
          "grantType"
        ],
        "environments": [
          "cff48ea9f6d8463a89a4152f4cb9b63a",
          "1954177776e54dd183e0e5736aeccedc"
        ],
        "apiScopes": [
          "3fcb5372de9547a88fba0590f6f3262d",
          "cfbfd1e49c894eef89c1cca44b91183a"
        ],
        "identityScopes": [
          "289e4b72eec44a7ead356a50f99070e2",
          "3fb18661a0b243b0abd834e738fe07ae"
        ],
        "ipAddressFilter": {
          "policy": "INTERNAL",
          "warnOnly": false,
          "allowList": []
        }
      }
    }
  }
}
